Add support for scale-as-tensor in quant ops

wenbinlyuTT opened the pull request:
### Ticket
#15934

### Problem description
Allowing the user to pass the quantization scale as a tensor.
This is a small functional addition, but it involves a few bug fixes in BinaryNg and warrants extensive unit testing.
This also paves the way to supporting per-channel quantization, in which both scale and zero-point must be passed as tensors.

### What's changed
Allowing the user to pass the quantization scale as a scalar-tensor by modifying the existing API entries to use `std::variant`.
Fix the incorrect assumption in BinaryNg that tensor A and scalar B always have the same data type/format, as well as the messy logic around data types/formats.
When the scale is passed as a scalar, force it to be f32 to improve accuracy and bypass #19453
When the scale is passed as a tensor, switch to a composite op fallback (accuracy and performance are not as good, and must workaround issues like #18955)
Add more quantization-related unit tests and parameter combinations, also add new assertions that catch cases missed by PCC (output tensor is uniform).

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

wenbinlyuTT committed:
Add support for scale-as-tensor in quant ops

Signed-off-by: wenbinlyuTT <wenbinlyu@tenstorrent.com>

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp:
```diff
@@ -399,9 +400,9 @@ BinaryNgDeviceOperation::ProgramFactory::cached_program_t BinaryNgDeviceOperatio
-    auto is_sfpu_op = operation_attributes.is_sfpu;
+    const bool is_sfpu_op = operation_attributes.is_sfpu;
+    const bool is_quant_op = operation_attributes.is_quant_op;
+    TT_FATAL(is_quant_op ? operation_attributes.is_sfpu : true, "Quantization op is SFPU-only");
```
```suggestion
    if (is_quant_op) {
        TT_FATAL(is_sfpu, "Quantization op is SFPU-only");
    }
```

sjameelTT commented on ttnn/cpp/ttnn/operations/eltwise/quantization/quantization.cpp:
```diff
@@ -30,17 +35,45 @@ Tensor QuantOp::invoke(
+    TT_FATAL(scale.get_logical_volume() == 1u, "Per-tensor quantize only takes a single scale value");
+
+    // The alternative is to pass ttnn::reciprocal(scale) to binary_ng, but it's expensive
+    const float scale_scalar = [](const DataType b_dtype, const Tensor& scale) {
```
Would this be extracting the scale on host? Isn't it preferable to read in the tensor in your kernel? At least for now this is okay, but long term you may want to do it within the kernel.

wenbinlyuTT replied:
Could you elaborate a bit on "read the tensor in your kernel"? Due to LLK restrictions we must use the reciprocal of the input scale tensor, is there any other way to do it on the device other than calling `ttnn::reciprocal`? I've benchmarked, the additional op approach is much more expensive than just fetching the scalar from the device and do reciprocal on the host. After all it's just a single scalar.

patrickroberts replied:
@wenbinlyuTT I think it could be achieved using `input_b_tensor_activations` (currently `rhs_activations` on `main`) containing a `UnaryWithParam` of `rdiv` with a value of `1`, and passing the `scale` as a `Tensor` to the primitive op rather than as a scalar. In that version of the kernel, the scalar value would be read from the CB, the reciprocal taken, and then passed as `b` to the binary op.

sjameelTT replied:
I am not sure about the internals of ttnn::reciprocal so you know better there. One concern I have is that in an actual model, moving the scalar onto host will break trace. Generally, we get a good performance boost from trace so it might be worth testing with that on if you haven't already.

Is it not possible to do:

noc_async_read(scalar_buffer_address...)
// Read in value into variable from L1
then just get your scalar value by doing 1/value in L1?

I would be surprised if this performs worse, though it may be a pain to fork the OP for just this.

wenbinlyuTT replied:
@sjameelTT Thanks, I understand now, that's totally doable and pretty neat, but yeah that change will be too specific for quant ops. I think we should focus on functional correctness first and worry about optimizations later.
I have switched to composite ops to deal with tensor inputs and no `to_vector` is used anymore.

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_utils.cpp:
```diff
@@ -331,15 +331,35 @@ void add_activation_defines(
+    } else if (dtype == DataType::INT32) {
         return std::bit_cast<uint32_t>(static_cast<int32_t>(scalar));
+    } else {
+        return pack_two_bfloat16_into_uint32({scalar, scalar});
```
Rather than using `else`, just use "early return" pattern:
```cpp
    if (dtype == DataType::FLOAT32 or is_quant_op) {
        return std::bit_cast<uint32_t>(scalar);
    }

    if (dtype == DataType::INT32) {
        return std::bit_cast<uint32_t>(static_cast<int32_t>(scalar));
    }

     return pack_two_bfloat16_into_uint32({scalar, scalar});
```

wenbinlyuTT replied:
Done

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/quantization/quantization.cpp:
```diff
@@ -89,8 +242,65 @@ Tensor DequantOp::invoke(
     const ttnn::DataType a_dtype = input_tensor.get_dtype();
-    const bool typecast_a = needs_typecast_to_bfloat16(a_dtype);
-    Tensor input_a = typecast_a ? typecast_to(DataType::BFLOAT16, input_tensor) : input_tensor;
+    ttnn::DataType c_dtype = ttnn::DataType::BFLOAT16;
+    if (output_dtype.has_value() && optional_output_tensor.has_value()) {
+        TT_FATAL(
+            output_dtype.value() == optional_output_tensor->dtype(),
+            "Mismatching output_dtype and output tensor dtype");
+        c_dtype = output_dtype.value();
+    } else if (output_dtype.has_value()) {
+        c_dtype = output_dtype.value();
+    } else if (optional_output_tensor.has_value()) {
+        c_dtype = optional_output_tensor->dtype();
+    }
```
This looks like it should be a function.

wenbinlyuTT replied:
Done

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/quantization/quantization.cpp:
```diff
@@ -89,8 +242,65 @@ Tensor DequantOp::invoke(
+    tt::stl::Span<const ttnn::operations::unary::UnaryWithParam> rhs_activations{};
+
+    // LLK dequant kernel does addition, so we need to negate zero_point
+    std::array<const ttnn::operations::unary::UnaryWithParam, 1> post_activations{
```
It is atypical for `std::array` to be passed a `const`-qualified value type, instead the array itself should be `const`-qualified. But starting from C++17 you actually don't need these template parameters at all though and can use CTAD (constructor template argument deduction) instead like this:
```suggestion
    const std::array post_activations{
```

wenbinlyuTT replied:
Done

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/quantization/quantization.hpp:
```diff
@@ -34,6 +44,42 @@ struct RequantOp {
         const std::optional<MemoryConfig>& memory_config = std::nullopt,
         std::optional<Tensor> optional_output_tensor = std::nullopt);
+
+    static Tensor invoke(
+        QueueId queue_id,
+        const Tensor& input_tensor,
+        const Tensor& in_scale,
+        const int32_t in_zero_point,
+        const float out_scale,
+        const int32_t out_zero_point,
+        const std::optional<int32_t> axis,
+        const std::optional<const DataType>& output_dtype = std::nullopt,
+        const std::optional<MemoryConfig>& memory_config = std::nullopt,
+        std::optional<Tensor> optional_output_tensor = std::nullopt);
+
+    static Tensor invoke(
+        QueueId queue_id,
+        const Tensor& input_tensor,
+        const float in_scale,
+        const int32_t in_zero_point,
+        const Tensor& out_scale,
+        const int32_t out_zero_point,
+        const std::optional<int32_t> axis,
+        const std::optional<const DataType>& output_dtype = std::nullopt,
+        const std::optional<MemoryConfig>& memory_config = std::nullopt,
+        std::optional<Tensor> optional_output_tensor = std::nullopt);
+
+    static Tensor invoke(
+        QueueId queue_id,
+        const Tensor& input_tensor,
+        const Tensor& in_scale,
+        const int32_t in_zero_point,
+        const Tensor& out_scale,
+        const int32_t out_zero_point,
+        const std::optional<int32_t> axis,
+        const std::optional<const DataType>& output_dtype = std::nullopt,
+        const std::optional<MemoryConfig>& memory_config = std::nullopt,
+        std::optional<Tensor> optional_output_tensor = std::nullopt);
```
So tempted to suggest `const std::variant<float, Tensor>& in_scale` and `const std::variant<float, Tensor>& out_scale` and condense this back to one overload. `std::visit` can actually be invoked with multiple variants so you don't even need to nest callbacks to do this either. That would be my personal recommendation, but I don't know how others feel about this suggestion, so don't do it if you don't feel strongly about it.

wenbinlyuTT replied:
I was thinking the same, especially since I'm going to add zero-point-as-tensors next and that's just going to add more combinations.

wenbinlyuTT replied:
Done

patrickroberts commented on ttnn/cpp/ttnn/operations/eltwise/quantization/quantization.cpp:
```diff
@@ -30,17 +35,45 @@ Tensor QuantOp::invoke(
+    TT_FATAL(scale.get_logical_volume() == 1u, "Per-tensor quantize only takes a single scale value");
+
+    // The alternative is to pass ttnn::reciprocal(scale) to binary_ng, but it's expensive
+    const float scale_scalar = [](const DataType b_dtype, const Tensor& scale) {
+        if (b_dtype == DataType::FLOAT32) {
+            return scale.to_vector<float>()[0];
+        } else {
+            return scale.to_vector<bfloat16>()[0].to_float();
+        }
+    }(b_dtype, scale);
```
```suggestion
    const auto scale_scalar = b_dtype == DataType::FLOAT32 ?
        scale.to_vector<float>()[0] :
        scale.to_vector<bfloat16>()[0].to_float();
```
I assume you've probably seen this immediately-invoked lambda pattern somewhere else, but it's only useful if you need `if constexpr` so the return type can be dependent and ill-formed instantiations can be discarded.

wenbinlyuTT replied:
Done

wenbinlyuTT committed:
Address some easier review comments

Signed-off-by: wenbinlyuTT <wenbinlyu@tenstorrent.com>

wenbinlyuTT commented on ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp:
```diff
@@ -535,27 +540,11 @@ BinaryNgDeviceOperation::ProgramFactory::cached_program_t BinaryNgDeviceOperatio
-    auto reader_defines = dataflow_defines;
+    // READER KERNEL
+    std::map<std::string, std::string> reader_defines;
+    add_dataflow_defines(reader_defines, a_dtype, is_sfpu_op);
```
Done

wenbinlyuTT committed:
Use variants for quant and dequant

Signed-off-by: wenbinlyuTT <wenbinlyu@tenstorrent.com>

wenbinlyuTT committed:
Switch requant to variants and add composite op fallback

Signed-off-by: wenbinlyuTT <wenbinlyu@tenstorrent.com>

wenbinlyuTT commented:
@patrickroberts I have switched everything to `std::variant` and implemented composite op fallbacks, please take a look.
The monstrosity in requant is the best I can do atm, `i32 .* f32 -> i32` just won't work because:
1. The activation trick on `ttnn::multiply` didn't work and I also rediscovered #18955
2. The activation trick on BinaryNg::MUL failed as well, probably due to https://github.com/tenstorrent/tt-metal/blob/2ad6823ae35af754520e70c76b5771f762f0e110/ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp#L34
3. Using `ttnn::typecast` twice gave me severe PCC issues, on top of the already low numerical accuracy of the composite op

I did a work around by calling dequant for this, because if zero-point is zero, dequant is infact doing `i32 .* f32 -> i32` on the SFPU...

dchenTT commented on ttnn/cpp/ttnn/operations/eltwise/quantization/quantization.cpp:
```diff
@@ -3,113 +3,247 @@
+    const std::optional<ttnn::Tensor>& output_tensor,
+    const ttnn::DataType default_dtype) {
+    if (output_dtype.has_value() && output_tensor.has_value()) {
+        TT_FATAL(output_dtype.value() == output_tensor->dtype(), "Mismatching output_dtype and output tensor dtype");
```
I think the guideline suggests to print the values. https://github.com/tenstorrent/tt-metal/blob/main/contributing/ErrorMessageBestPractices.md

wenbinlyuTT replied:
@dchenTT OK, I will fix this in a follow-up PR
